res/controllers/Traktor-Kontrol-S2-MK1-hid-scripts.js

```diff
@@ -1,1364 +1,1525 @@
+                key: "button_parameter" + this.number,
+            };
         }
-    } else {
-        if (TraktorS2MK1.leftEncoderPressed[field.group]) {
-            var beatjumpSize = engine.getValue(field.group, "beatjump_size");
-            if (delta === 1) {
-                beatjumpSize *= 2;
-            } else {
-                beatjumpSize /= 2;
-            }
-            engine.setValue(field.group, "beatjump_size", beatjumpSize);
+    }
+    getKnobGroupAndKey() {
+        const focusedEffect = engine.getValue(this.group, "focused_effect");
+        if (focusedEffect === 0) {
+            return {
+                group: this.groupPrefix + "_Effect" + this.number + "]",
+                key: "meta",
+            };
         } else {
-            if (delta === 1) {
-                script.triggerControl(field.group, "beatjump_forward");
-            } else {
-                script.triggerControl(field.group, "beatjump_backward");
-            }
+            return {
+                group: this.groupPrefix + "_Effect" + focusedEffect + "]",
+                key: "parameter" + this.number,
+            };
         }
     }
-};
+    ledCallbackNormal(value) {
+        this.ledCallback(value === 1);
+    }
+    ledCallbackFocused(value) {
+        this.ledCallback(value === this.number);
+    }
+    ledCallback(value) {
+        this.controller.setOutput(this.group, "!effectbutton" + this.number,
+            value ? ButtonBrightnessOn : ButtonBrightnessOff, !this.effectUnit.parent.batchingLEDUpdate);
+    }
+}
+
+class TraktorS2MK1Class {
+    constructor() {
+        this.controller = new HIDController();
+
+        // When true, packets will not be sent to the controller.
+        // Used when updating multiple LEDs simultaneously.
+        this.batchingLEDUpdate = false;
+
+        this.browseEncoder = new Encoder();
+
+        this.decks = [
+            new DeckClass(this, 1),
+            new DeckClass(this, 2),
+        ];
+        this.effectUnits = [
+            new EffectUnit(this, 1),
+            new EffectUnit(this, 2),
+        ];
+        this.rawCalibration = {};
+        this.calibration = null;
+    }
+    registerInputPackets() {
+        // Values in input report 0x01 are all buttons, except the jog wheels.
+        // An exclamation point indicates a specially-handled function.  Everything else is a standard
+        // Mixxx control object name.
+        const InputReport0x01 = new HIDPacket("InputReport0x01", 0x01, this.inputReport0x01Callback.bind(this));
+        // Most items in the input report 0x02 are controls that go from 0-4095.
+        // There are also some 4 bit encoders.
+        const InputReport0x02 = new HIDPacket("InputReport0x02", 0x02, this.inputReport0x02Callback.bind(this));
+
+        this.decks[0].registerInputs(InputReport0x01, InputReport0x02, {
+            gainEncoderPress: [0x0E, 0x01],
+            shift: [0x0D, 0x80],
+            sync: [0x0D, 0x40],
```

Comment:
This registerInputs structure is very difficult to understand. For me it's not clear anymore, in which InputReport an input data item is located. Please rethink how to make this part of the code easier readable for others.

Reply:
Something like this?